File: clsMauroConnection.vb
Imports System.Collections.Generic
Imports System.Net.Http
Imports System.Net.Http.Headers
Imports System.Text.Json

Imports System.Text.Json.Serialization
Imports System.Text
Imports System.IO
Imports System.Xml

Namespace MauroAPI
    Class clsMauroConnection
        Private Shared ReadOnly client As HttpClient = New HttpClient()
        Public Shared Property Username As String
        Public Shared Property Password As String
        Public Shared Property Endpoint As String = "https://modelcatalogue.cs.ox.ac.uk/nhsd"

        Public Shared Property LoginStatus As Boolean = False

        ' Set up basic strings for the API
        Const LoginAPI = "/api/authentication/login"
        Const LogoutAPI = "/api/authentication/logout"
        Const ModelAPI = "/api/dataModels"
        Const JSONMime = "application/json"

        Public Shared Property LoginDetails As LoginResponse = Nothing

        Public Shared Property MauroModels As Models

        Public Shared Async Function Login() As Task(Of LoginResponse)
            client.DefaultRequestHeaders.Accept.Clear()
            client.DefaultRequestHeaders.Accept.Add(New MediaTypeWithQualityHeaderValue(JSONMime))

            client.DefaultRequestHeaders.Add("User-Agent", ".NET Mauro Template Serialiser")
            Dim MauroLogin As New LoginRequest With {.username = Username, .password = Password}
            Dim strBody = JsonSerializer.Serialize(MauroLogin)
            Dim Body As New StringContent(strBody, Encoding.UTF8, "application/json")

            Dim PostResult = Await client.PostAsync(Endpoint & LoginAPI, Body)

            Dim JSONResult = Await PostResult.Content.ReadAsStreamAsync

            LoginDetails = Await JsonSerializer.DeserializeAsync(Of LoginResponse)(JSONResult)
            LoginStatus = True
            Return LoginDetails
        End Function

        Public Shared Async Function GetModels() As Task(Of Models)
            Dim JSONString = Await GetMauroJSON(ModelAPI)
            MauroModels = JsonSerializer.Deserialize(Of Models)(JSONString)
            Return MauroModels
        End Function

        Public Shared Async Function GetModel(ID As Guid) As Task(Of Model)
            Dim JSONString = Await GetMauroJSON(ModelAPI & "/" & ID.ToString)
            Dim MauroModel = JsonSerializer.Deserialize(Of Model)(JSONString)
            Return MauroModel
        End Function



        Private Shared Async Function GetMauroJSON(API As String) As Task(Of String)

            If LoginStatus = False Then
                Await Login()
            End If
            ' Async retrieval of Mauro API JSON
            client.DefaultRequestHeaders.Accept.Clear()
            client.DefaultRequestHeaders.Accept.Add(New MediaTypeWithQualityHeaderValue(JSONMime))
            client.DefaultRequestHeaders.Add("User-Agent", ".NET Mauro Template Serialiser")

            ' Retrieve the JSON from the Mauro endpoint
            Dim GetResult = Await client.GetAsync(Endpoint & API)
            Dim JSONResult = Await GetResult.Content.ReadAsStreamAsync
            Dim reader As New StreamReader(JSONResult)

            Dim JSONstring = reader.ReadToEnd()
            Return JSONstring
        End Function

        Public Shared Async Function Logout() As Task
            client.DefaultRequestHeaders.Accept.Add(New MediaTypeWithQualityHeaderValue(JSONMime))

            client.DefaultRequestHeaders.Add("User-Agent", ".NET Mauro Template Serialiser")

            Dim PostResult = Await client.GetAsync(Endpoint & LogoutAPI)

            Dim ResultContent = Await PostResult.Content.ReadAsStringAsync
            Console.WriteLine(ResultContent)
        End Function

    End Class

    Public Class LoginResponse

        <JsonPropertyName("id")>
        Public Property id As Guid
        <JsonPropertyName("emailAddress")>
        Public Property emailAddress As String

        <JsonPropertyName("firstName")>
        Public Property firstName As String
        <JsonPropertyName("lastName")>
        Public Property lastName As String

        <JsonPropertyName("userRole")>
        Public Property userRole As String
        <JsonPropertyName("disabled")>
        Public Property disabled As Boolean
    End Class
    ''' <summary>
    ''' Mauro dataModels data structure
    ''' count of models as integer
    ''' list of Mauro model data structures
    ''' </summary>
    Public Class Models
        <JsonPropertyName("count")>
        Public Property count As Integer

        <JsonPropertyName("items")>
        Public Property items As List(Of Model)
    End Class
    ''' <summary>
    ''' Mauro dataModel datastructure
    ''' </summary>
    Public Class Model

        <JsonPropertyName("id")>
        Public Property id As Guid

        <JsonPropertyName("domainType")>
        Public Property domainType As String

        <JsonPropertyName("label")>
        Public Property label As String

        <JsonPropertyName("aliases")>
        Public Property aliases As Object ' List(Of String)

        <JsonPropertyName("description")>
        Public Property description As String

        <JsonPropertyName("author")>
        Public Property author As String

        <JsonPropertyName("organisation")>
        Public Property organisation As String

        <JsonPropertyName("editable")>
        Public Property editable As Boolean

        <JsonPropertyName("documentationVersion")>
        Public Property documentationVersion As String

        <JsonPropertyName("lastUpdated")>
        Public Property lastUpdated As Date

        <JsonPropertyName("classifiers")>
        Public Property classifiers() As List(Of Classifier)

        <JsonPropertyName("type")>
        Public Property type As String

        <JsonPropertyName("finalised")>
        Public Property finalised As Boolean

        Public Property metadata As List(Of Metadata)

    End Class

    Public Class Classifier
        <JsonPropertyName("id")>
        Public Property id As Guid

        <JsonPropertyName("label")>
        Public Property label As String
        <JsonPropertyName("lastUpdated")>
        Public Property lastUpdated As Date
    End Class

    Public Class Metadata
        '        {
        '    "id": "c9a36d30-2c6a-4dd0-a792-a337a2eca9c8",
        '    "namespace": "ox.softeng.metadatacatalogue.dataloaders.hdf",
        '    "key": "Volumes",
        '    "value": "Varies annually: in 2013/14, 18.2m finished consultant episodes (FCEs) and 15.5m Finished Admission Episodes (FAEs)",
        '    "lastUpdated": "2019-10-03T09:15:12.082Z"
        '}

        <JsonPropertyName("id")>
        Public Property id As Guid

        <JsonPropertyName("namespace")>
        Public Property mauroNamespace As String

        <JsonPropertyName("key")>
        Public Property key As String

        <JsonPropertyName("value")>
        Public Property value As String
        <JsonPropertyName("lastUpdated")>
        Public Property lastUpdated As Date

    End Class

    Public Class LoginRequest
        Property username As String
        Property password As String
    End Class
End Namespace

File: clsMauroConnection.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace MauroTemplate.MauroAPI
{
    class clsMauroConnection
    {
        private static readonly HttpClient client = new HttpClient();

        public static string Username { get; set; }
        public static string Password { get; set; }
        public static string Endpoint { get; set; } = "https://modelcatalogue.cs.ox.ac.uk/nhsd";
        public static bool LoginStatus { get; set; } = false;

        // Set up basic strings for the API
        private const string LoginAPI = "/api/authentication/login";
        private const string LogoutAPI = "/api/authentication/logout";
        private const string ModelAPI = "/api/dataModels";
        private const string JSONMime = "application/json";

        public static LoginResponse LoginDetails { get; set; } = null;
        public static Models MauroModels { get; set; }

        public static async Task<LoginResponse> Login()
        {
            client.DefaultRequestHeaders.Accept.Clear();
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(JSONMime));
            client.DefaultRequestHeaders.Add("User-Agent", ".NET Mauro Template Serialiser");
            var MauroLogin = new LoginRequest() { username = Username, password = Password };
            string strBody = JsonSerializer.Serialize(MauroLogin);
            var Body = new StringContent(strBody, Encoding.UTF8, "application/json");
            var PostResult = await client.PostAsync(Endpoint + LoginAPI, Body);
            var JSONResult = await PostResult.Content.ReadAsStreamAsync();
            LoginDetails = await JsonSerializer.DeserializeAsync<LoginResponse>(JSONResult);
            LoginStatus = true;
            return LoginDetails;
        }

        public static async Task<Models> GetModels()
        {
            string JSONString = await GetMauroJSON(ModelAPI);
            MauroModels = JsonSerializer.Deserialize<Models>(JSONString);
            return MauroModels;
        }

        public static async Task<Model> GetModel(Guid ID)
        {
            string JSONString = await GetMauroJSON(ModelAPI + "/" + ID.ToString());
            var MauroModel = JsonSerializer.Deserialize<Model>(JSONString);
            return MauroModel;
        }

        private static async Task<string> GetMauroJSON(string API)
        {
            if (LoginStatus == false)
            {
                await Login();
            }
            // Async retrieval of Mauro API JSON
            client.DefaultRequestHeaders.Accept.Clear();
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(JSONMime));
            client.DefaultRequestHeaders.Add("User-Agent", ".NET Mauro Template Serialiser");

            // Retrieve the JSON from the Mauro endpoint
            var GetResult = await client.GetAsync(Endpoint + API);
            var JSONResult = await GetResult.Content.ReadAsStreamAsync();
            var reader = new StreamReader(JSONResult);
            string JSONstring = reader.ReadToEnd();
            return JSONstring;
        }

        public static async Task Logout()
        {
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(JSONMime));
            client.DefaultRequestHeaders.Add("User-Agent", ".NET Mauro Template Serialiser");
            var PostResult = await client.GetAsync(Endpoint + LogoutAPI);
            string ResultContent = await PostResult.Content.ReadAsStringAsync();
            Console.WriteLine(ResultContent);
        }
    }

    public class LoginResponse
    {
        [JsonPropertyName("id")]
        public Guid id { get; set; }
        [JsonPropertyName("emailAddress")]
        public string emailAddress { get; set; }
        [JsonPropertyName("firstName")]
        public string firstName { get; set; }
        [JsonPropertyName("lastName")]
        public string lastName { get; set; }
        [JsonPropertyName("userRole")]
        public string userRole { get; set; }
        [JsonPropertyName("disabled")]
        public bool disabled { get; set; }
    }
    /// <summary>
    /// Mauro dataModels data structure
    /// count of models as integer
    /// list of Mauro model data structures
    /// </summary>
    public class Models
    {
        [JsonPropertyName("count")]
        public int count { get; set; }
        [JsonPropertyName("items")]
        public List<Model> items { get; set; }
    }
    /// <summary>
    /// Mauro dataModel datastructure
    /// </summary>
    public class Model
    {
        [JsonPropertyName("id")]
        public Guid id { get; set; }
        [JsonPropertyName("domainType")]
        public string domainType { get; set; }
        [JsonPropertyName("label")]
        public string label { get; set; }
        [JsonPropertyName("aliases")]
        public object aliases { get; set; } // List(Of String)
        [JsonPropertyName("description")]
        public string description { get; set; }
        [JsonPropertyName("author")]
        public string author { get; set; }
        [JsonPropertyName("organisation")]
        public string organisation { get; set; }
        [JsonPropertyName("editable")]
        public bool editable { get; set; }
        [JsonPropertyName("documentationVersion")]
        public string documentationVersion { get; set; }
        [JsonPropertyName("lastUpdated")]
        public DateTime lastUpdated { get; set; }
        [JsonPropertyName("classifiers")]
        public List<Classifier> classifiers { get; set; }
        [JsonPropertyName("type")]
        public string type { get; set; }
        [JsonPropertyName("finalised")]
        public bool finalised { get; set; }
        public List<Metadata> metadata { get; set; }
    }

    public class Classifier
    {
        [JsonPropertyName("id")]
        public Guid id { get; set; }
        [JsonPropertyName("label")]
        public string label { get; set; }
        [JsonPropertyName("lastUpdated")]
        public DateTime lastUpdated { get; set; }
    }

    public class Metadata
    {
        // {
        // "id": "c9a36d30-2c6a-4dd0-a792-a337a2eca9c8",
        // "namespace": "ox.softeng.metadatacatalogue.dataloaders.hdf",
        // "key": "Volumes",
        // "value": "Varies annually: in 2013/14, 18.2m finished consultant episodes (FCEs) and 15.5m Finished Admission Episodes (FAEs)",
        // "lastUpdated": "2019-10-03T09:15:12.082Z"
        // }

        [JsonPropertyName("id")]
        public Guid id { get; set; }
        [JsonPropertyName("namespace")]
        public string mauroNamespace { get; set; }
        [JsonPropertyName("key")]
        public string key { get; set; }
        [JsonPropertyName("value")]
        public string value { get; set; }
        [JsonPropertyName("lastUpdated")]
        public DateTime lastUpdated { get; set; }
    }

    public class LoginRequest
    {
        public string username { get; set; }
        public string password { get; set; }
    }
}